### PR TITLE
Fix double agent messages with rate limit errors

### PR DESCRIPTION
**Pull request by pkukielka**
Fixes https://github.com/sourcegraph/jetbrains/issues/233

## Test plan

Automatic tests to cover the issue were added. 

## Details

Error `Error: Cannot add a bot message after a bot message` was popping out when message was sent to agent with rate limit exhausted. It was caused by `SimpleChatPanelProvider` closing  `typewriter`, which in turn was calling `close` callback from `sendLLMRequest` in `SimpleChatPanelProvider`.
Close callback was then dumping remaining content, resulting in adding agent message twice.

```
    at SimpleChatPanelProvider.addBotMessageWithGuardrails (cody/vscode/src/chat/chat-view/SimpleChatPanelProvider.ts:855:21)
    at Object.close (cody/vscode/src/chat/chat-view/SimpleChatPanelProvider.ts:690:26)
    at Object.close (cody/vscode/src/chat/chat-view/SimpleChatPanelProvider.ts:729:27)
    at Typewriter.stop (cody/lib/shared/src/chat/typewriter.ts:116:27)
    at Object.onError (cody/vscode/src/chat/chat-view/SimpleChatPanelProvider.ts:748:32)
    at onErrorOnce (cody/lib/shared/src/sourcegraph-api/completions/nodeClient.ts:29:20)
    at handleError (cody/lib/shared/src/sourcegraph-api/completions/nodeClient.ts:82:25)
    at IncomingMessage.<anonymous> (cody/lib/shared/src/sourcegraph-api/completions/nodeClient.ts:107:41)
    at IncomingMessage.emit (node:events:529:35)
    at IncomingMessage.emit (node:domain:489:12)
```

### CODE REVIEW

**Review by abeatrix — APPROVED** [review]
lgtm!

**Review by olafurpg — APPROVED** [review]
LGTM 👍🏻 Nice job tracking this down and good to have a test case to reproduce it.

**Comment by olafurpg** [comment]
> window.showErrorMessage: Failed to download bfg from URL https://github.com/sourcegraph/bfg/releases/download/v5.2.12792/bfg-linux-x64.zip: PollyError: [Polly] [adapter:node-http] Recording for the following request is not found and `recordIfMissing` is `false`.

I'm surprised to see this error since we configure Polly to passthrough requests to github.com

https://sourcegraph.com/github.com/sourcegraph/cody@olafurpg/evaluate-more/-/blob/agent/src/cli/jsonrpc.ts?L162

The fact this only fails in Node v16 indicates something else might be going on...

File: lib/shared/src/chat/typewriter.ts [hunk 1]
```diff
@@ -113,9 +118,16 @@ export class Typewriter implements IncrementalTextConsumer {
         }
         // Clean up the consumer, finished promise.
         if (this.upstreamClosed) {
-            this.consumer.close()
-        } else {
-            console.error('Typewriter stopped', error)
+            if (error) {
```
Comment (abeatrix) [inline lib/shared/src/chat/typewriter.ts:121]:
```suggestion
            if (error) {
                console.error('Typewriter stopped', error)
                if (this.consumer.error) {
                    this.consumer.error(error)
                    return
                }
            }
            this.consumer.close()
```
would this do the same thing?

**Review by abeatrix — APPROVED** [review]
Confirmed the rate limit test is passing when merged with the latest changes in main:

image

The large chat test will need to be updated accordingly though. Thanks for all your help on this @pkukielka !

**Comment by olafurpg** [comment]
@pkukielka close this PR? My understanding is that this has been merged through other PRs

**Pull request closed without merging**